src/main/flight/imu.c

```diff
@@ -261,150 +243,158 @@ static void imuCheckAndResetOrientationQuaternion(const float ax, const float ay
 
-            ex = rMat[2][0] * ez_ef;
-            ey = rMat[2][1] * ez_ef;
-            ez = rMat[2][2] * ez_ef;
-        }
-        else {
-            ex = 0;
-            ey = 0;
-            ez = 0;
+            // Rotate error back into body frame
+            rotateVectorByQuaternion(&vErr, &vErr, &orientation);
         }
 
         // Compute and apply integral feedback if enabled
         if (imuRuntimeConfig.dcm_ki_mag > 0.0f) {
             // Stop integrating if spinning beyond the certain limit
             if (spin_rate_sq < sq(DEGREES_TO_RADIANS(SPIN_RATE_LIMIT))) {
-                integralMagX += imuRuntimeConfig.dcm_ki_mag * ex * dt;    // integral error scaled by Ki
-                integralMagY += imuRuntimeConfig.dcm_ki_mag * ey * dt;
-                integralMagZ += imuRuntimeConfig.dcm_ki_mag * ez * dt;
+                t_fp_vector vTmp;
+
+                // integral error scaled by Ki
+                vectorScale(&vTmp, &vErr, imuRuntimeConfig.dcm_ki_mag * dt);
+                vectorAdd(&vIntegralMag, &vIntegralMag, &vTmp);
 
-                gx += integralMagX;
-                gy += integralMagY;
-                gz += integralMagZ;
+                // apply integral feedback
+                vectorAdd(&vRotation, &vRotation, &vIntegralMag);
             }
         }
 
         // Calculate kP gain and apply proportional feedback
-        gx += kpMag * ex;
-        gy += kpMag * ey;
-        gz += kpMag * ez;
+        vectorScale(&vErr, &vErr, imuRuntimeConfig.dcm_kp_mag * imuGetPGainScaleFactor());
+        vectorAdd(&vRotation, &vRotation, &vErr);
     }
 
 
     /* Step 2: Roll and pitch correction -  use measured acceleration vector */
     if (useAcc) {
-        float kpAcc = imuRuntimeConfig.dcm_kp_acc * imuGetPGainScaleFactor();
-        const float accRecipNorm = invSqrt(ax * ax + ay * ay + az * az);
+        static const t_fp_vector vGravity = { .V = { 0.0f, 0.0f, 1.0f } };
+        t_fp_vector vEstGravity, vAcc, vErr;
 
-        // Just scale by 1G length - That's our vector adjustment. Rather than
-        // using one-over-exact length (which needs a costly square root), we already
-        // know the vector is enough "roughly unit length" and since it is only weighted
-        // in by a certain amount anyway later, having that exact is meaningless. (c) MasterZap
-        ax *= accRecipNorm;
-        ay *= accRecipNorm;
-        az *= accRecipNorm;
+        // Calculate estimated gravity vector in body frame
+        rotateVectorByQuaternion(&vEstGravity, &vGravity, &orientation);    // EF -> BF
 
         // Error is sum of cross product between estimated direction and measured direction of gravity
-        ex = (ay * rMat[2][2] - az * rMat[2][1]);
-        ey = (az * rMat[2][0] - ax * rMat[2][2]);
-        ez = (ax * rMat[2][1] - ay * rMat[2][0]);
+        vectorNormalize(&vAcc, accBF);
+        vectorCrossProduct(&vErr, &vAcc, &vEstGravity);
```

Comment:
Possible improvement:
Cross product will be small when vectors are close to antiparallel (initial convergence problem). Using dot product &vAcc, &vEstGravity, it is easy to detect large error and update error estimation accordingly (for example normalize vErr when dot(&vAcc, &vEstGravity) < 0

Reply:
I'll try that

Reply:
It will usually apply only on initial guess. But it is also metric to detect 'large deviation' that is done explicitly to reset orientation in iNav ...